Nomad client fails to start a job

### Nomad version
```
Installed from official RPM repo

Nomad v1.5.6
BuildDate 2023-05-19T18:26:13Z
Revision 8af70885c02ab921dedbdf6bc406a1e886866f80
```

### Operating system and Environment details
```
$ cat /etc/os-release 
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

$ uname -a
Linux hostname.redacted 3.10.0-123.20.1.el7.x86_64 #1 SMP Thu Jan 29 18:05:33 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux

$ cat /proc/cmdline 
BOOT_IMAGE=/vmlinuz-3.10.0-123.20.1.el7.x86_64 root=/dev/mapper/VolGroup-lv_root ro nomodeset rd_NO_PLYMOUTH rd.lvm.lv=VolGroup/lv_swap vconsole.keymap=us plymouth.enable=0 rd.lvm.lv=VolGroup/lv_root crashkernel=auto vconsole.font=latarcyrheb-sun16 rhgb quiet LANG=en_US.UTF-8 rcu_nocbs=1,3 nohz_full=1,3 intel_pstate=disable

$ cat /etc/nomad.d/nomad.json 
{
  "data_dir": "/data/nomad",
  "bind_addr": "0.0.0.0",
  "datacenter": "perf",
  "region": "hk",
  "server": {
    "enabled": false,
  },
  "client": {
    "enabled": true,
    "chroot_env": {
      "/bin": "/bin",
      "/etc": "/etc",
      "/lib": "/lib",
      "/lib32": "/lib32",
      "/lib64": "/lib64",
      "/run/resolvconf": "/run/resolvconf",
      "/sbin": "/sbin",
      "/usr": "/usr"
    }
  },
  "consul": {
    "server_service_name": "nomad-server",
    "client_service_name": "nomad-client",
    "address": "127.0.0.1:8500",
    "auto_advertise": true,
    "client_auto_join": true,
    "server_auto_join": true
  }
}
```

### Issue
Submitting job fails with:

```
Jun 07 12:48:08 hostname.redacted nomad[7485]: 2023-06-07T12:48:08.590+0800 [INFO]  client.alloc_runner.task_runner: Task event: alloc_id=c6998dae-aa8a-f030-462b-e31aa2db2233 task=sleep type="Driver Failure" msg="failed to launch command with executor: rpc error: code = Unknown desc = unable to start container process: error during container init: error setting cgroup config for procHooks process: failed to write \"0\": write /sys/fs/cgroup/memory/nomad/2f4cdd07-1faa-9b22-178c-b5c9b1dadafd/memory.swappiness: invalid argument" failed=false
Jun 07 12:48:08 hostname.redacted nomad[7485]: 2023-06-07T12:48:08.692+0800 [ERROR] client.alloc_runner.task_runner: running driver failed: alloc_id=c6998dae-aa8a-f030-462b-e31aa2db2233 task=sleep error="failed to launch command with executor: rpc error: code = Unknown desc = unable to start container process: error during container init: error setting cgroup config for procHooks process: failed to write \"0\": write /sys/fs/cgroup/memory/nomad/2f4cdd07-1faa-9b22-178c-b5c9b1dadafd/memory.swappiness: invalid argument"
```

### Reproduction steps
* Try to start a job in an env as described above

#### Expected Result
* Job started

#### Actual Result
* Job allocation fails

### Job file (if appropriate)
```
job "hello-world" {
  datacenters = ["perf"]
  group "servers" {
    task "sleep" {
      driver = "exec"
      config {
        command = "sleep"
        args    = ["infinity"]
      }
    }
  }
}
```

### Nomad Server logs (if appropriate)
N/A

### Nomad Client logs (if appropriate)
See above

--- a/.changelog/17625.txt
+++ b/.changelog/17625.txt
@@ -1,3 +1,3 @@
 ```release-note:bug
-client: Fixed a bug where Nomad incorrectly write to memory swappiness cgroup on old kernels
+client: Fixed a bug where Nomad incorrectly wrote to memory swappiness cgroup on old kernels
 ```

--- a/.changelog/17625.txt
+++ b/.changelog/17625.txt
@@ -1,0 +1,3 @@
+```release-note:bug
+client: Fixed a bug where Nomad incorrectly write to memory swappiness cgroup on old kernels
+```

--- a/client/lib/cgutil/cgutil_linux.go
+++ b/client/lib/cgutil/cgutil_linux.go
@@ -162,9 +162,11 @@ func MaybeDisableMemorySwappiness() *uint64 {
 		return zero
 	}
 
-	// cgroups v1 check if the root and swappiness interface exist
-	_, err := os.Stat("/sys/fs/cgroup/memory/memory.swappiness")
-	if os.IsNotExist(err) {
+	// cgroups v1 detect if swappiness is supported by attempting to write to
+	// the nomad parent cgroup swappiness interface
+	e := &editor{fromRoot: "memory/nomad"}
+	err := e.write("memory.swappiness", "0")
+	if err != nil {
 		return bypass
 	}
 

--- a/client/lib/cgutil/cgutil_linux.go
+++ b/client/lib/cgutil/cgutil_linux.go
@@ -11,6 +11,7 @@ import (
 	"path/filepath"
 
 	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/nomad/helper/pointer"
 	"github.com/hashicorp/nomad/helper/uuid"
 	"github.com/opencontainers/runc/libcontainer/cgroups"
 	lcc "github.com/opencontainers/runc/libcontainer/configs"
@@ -147,4 +148,25 @@ func CopyCpuset(source, destination string) error {
 	}
 
 	return nil
+}
+
+// MaybeDisableMemorySwappiness will disable memory swappiness, if that controller
+// is available. Always the case for cgroups v2, but is not always the case on
+// very old kernels with cgroups v1.
+func MaybeDisableMemorySwappiness() *uint64 {
+	bypass := (*uint64)(nil)
+	zero := pointer.Of[uint64](0)
+
+	// cgroups v2 always set zero
+	if UseV2 {
+		return zero
+	}
+
+	// cgroups v1 check if the root and swappiness interface exist
+	_, err := os.Stat("/sys/fs/cgroup/memory/memory.swappiness")
+	if os.IsNotExist(err) {
+		return bypass
+	}
+
+	return zero
 }

--- a/client/lib/cgutil/cgutil_linux_test.go
+++ b/client/lib/cgutil/cgutil_linux_test.go
@@ -130,3 +130,12 @@ func TestUtil_CopyCpuset(t *testing.T) {
 		require.Equal(t, "0-1", strings.TrimSpace(value))
 	})
 }
+
+func TestUtil_MaybeDisableMemorySwappiness(t *testing.T) {
+	ci.Parallel(t)
+
+	// will return 0 on any reasonable kernel (both cgroups v1 and v2)
+	value := MaybeDisableMemorySwappiness()
+	must.NotNil(t, value)
+	must.Eq(t, 0, *value)
+}

--- a/drivers/shared/executor/executor_linux.go
+++ b/drivers/shared/executor/executor_linux.go
@@ -689,9 +689,8 @@ func configureCgroups(cfg *lconfigs.Config, command *ExecCommand) error {
 		cfg.Cgroups.Resources.Memory = memHard * 1024 * 1024
 		cfg.Cgroups.Resources.MemoryReservation = memSoft * 1024 * 1024
 
-		// Disable swap to avoid issues on the machine
-		var memSwappiness uint64
-		cfg.Cgroups.Resources.MemorySwappiness = &memSwappiness
+		// Disable swap if possible, to avoid issues on the machine
+		cfg.Cgroups.Resources.MemorySwappiness = cgutil.MaybeDisableMemorySwappiness()
 	}
 
 	cpuShares := res.Cpu.CpuShares

